File: folkmq-lang-js/folkmq/src/common/MqUtils.ts
import {Message, MessageDefault} from "@noear/socket.d/transport/core/Message";
import {MqConstants} from "./MqConstants";
import {MqMetasResolverV2} from "./MqMetasResolverV2";
import {MqMetasResolverV1} from "./MqMetasResolverV1";
import {MqMetasResolver} from "./MqMetasResolver";
import {Session} from "@noear/socket.d/transport/core/Session";
import {MqMetasV2} from "./MqMetasV2";
import {MqMetasResolverV3} from "./MqMetasResolverV3";
import {Entity} from "@noear/socket.d/transport/core/Entity";

/**
 * 消息工具类
 *
 * @author noear
 * @since 1.0
 * @since 1.2
 */
export class MqUtils {
    private static v1 = new MqMetasResolverV1();
    private static v2 = new MqMetasResolverV2();
    private static v3 = new MqMetasResolverV3();

    static getLast(): MqMetasResolver {
        return MqUtils.v3;
    }

    static getOf(t: Session | Entity | null): MqMetasResolver {
        if (t == null) {
            return MqUtils.getLast();
        }

        if (t instanceof MessageDefault) {
            let m = t as Message;
            let ver = m.metaOrDefault(MqMetasV2.MQ_META_VID, "1");
            return MqUtils.resolve(ver);
        } else {
            let s = t as Session;
            let ver = s.handshake().paramOrDefault(MqConstants.FOLKMQ_VERSION, "1");
            return MqUtils.resolve(ver);
        }
    }

    static resolve(ver: string): MqMetasResolver {
        if ("1" == ver) {
            return MqUtils.v1;
        } else if ("2" == ver) {
            return MqUtils.v2;
        } else {
            return MqUtils.v3;
        }
    }
}

File: folkmq-lang-java/folkmq/src/main/java/org/noear/folkmq/common/MqUtils.java
package org.noear.folkmq.common;

import org.noear.socketd.transport.core.Entity;
import org.noear.socketd.transport.core.Session;

/**
 * 消息工具类
 *
 * @author noear
 * @since 1.0
 * @since 1.2
 */
public class MqUtils {
    private static MqMetasResolver v1 = new MqMetasResolverV1();
    private static MqMetasResolver v2 = new MqMetasResolverV2();
    private static MqMetasResolver v3 = new MqMetasResolverV3();

    public static MqMetasResolver getLast() {
        return v2;
    }

    public static MqMetasResolver getOf(Session s) {
        //def=1
        String ver = s.handshake().paramOrDefault(MqConstants.FOLKMQ_VERSION, "1");
        return resolve(ver);
    }

    public static MqMetasResolver getOf(Entity m) {
        if (m == null) {
            return getLast();
        }

        //def=1
        String ver = m.metaOrDefault(MqMetasV2.MQ_META_VID, "1");
        return resolve(ver);
    }

    private static MqMetasResolver resolve(String ver){
        if ("1".equals(ver)) {
            return v1;
        } else if ("2".equals(ver)) {
            return v2;
        } else {
            return v3;
        }
    }
}